ksql-rest-app/src/test/resources/ksql-plan-schema/schema-final.json

```diff
@@ -0,0 +1,885 @@
+{
+  "$schema" : "http://json-schema.org/draft-04/schema#",
+  "title" : "Ksql Plan",
+  "oneOf" : [ {
+    "$ref" : "#/definitions/KsqlPlanV1"
+  } ],
+  "definitions" : {
+    "KsqlPlanV1" : {
+      "type" : "object",
+      "additionalProperties" : false,
+      "properties" : {
+        "@type" : {
+          "type" : "string",
+          "enum" : [ "ksqlPlanV1" ],
+          "default" : "ksqlPlanV1"
+        },
+        "statementText" : {
+          "type" : "string"
+        },
+        "ddlCommand" : {
+          "oneOf" : [ {
+            "$ref" : "#/definitions/CreateStreamCommand"
+          }, {
+            "$ref" : "#/definitions/CreateTableCommand"
+          }, {
+            "$ref" : "#/definitions/RegisterTypeCommand"
+          }, {
+            "$ref" : "#/definitions/DropSourceCommand"
+          }, {
+            "$ref" : "#/definitions/DropTypeCommand"
+          } ]
+        },
+        "queryPlan" : {
+          "$ref" : "#/definitions/QueryPlan"
+        }
+      },
+      "title" : "ksqlPlanV1",
+      "required" : [ "@type", "statementText", "ddlCommand", "queryPlan" ]
+    },
+    "CreateStreamCommand" : {
+      "type" : "object",
+      "additionalProperties" : false,
+      "properties" : {
+        "@type" : {
+          "type" : "string",
+          "enum" : [ "createStreamV1" ],
+          "default" : "createStreamV1"
```

Comment:
> If that's the case, then we can drop this @type property as it will always be createStreamV1. Same for other objects.

No matter what we name the step schemas, we need a type property in the schema. The names in the JSON schema are just for resolving internal references within the schema spec (`$ref`s). JSON data has no type information. We need a type property so that the validator or deserializer can find the right schema or POJO type to validate against or deserialize to, respectively.

Reply:
So we need the `@type` for the Jackson stuff to work?  Shame.  That answers not this comment but [this one](https://github.com/confluentinc/ksql/pull/3969/files#r350645611).

This comment is more about two things:
1. I think the outer type should be named `CreateStreamCommandV1`, in the same way that its `KsqlPlanV1`, i.e. just making them consistent.  The reason for this is:
2. This `@type` shouldn't be an enum, it should be `createStreamV1`. Why? Well, if we were to release a v2 of `CreateStreamCommand` it will likely have a different set of properties than v1.   For example, let's say it doesn't have the `schema` property, but does have a new `awesomeness` property. With the current approach `CreateStreamCommand` would have both `schema` and `awesomeness` properties, but neither could be marked as `required` . Better to have two distinct, strongly typed / validated, objects, i.e. rather than having:

```json
{
"CreateStreamCommand" : {
      "type" : "object",
      "additionalProperties" : false,
      "properties" : {
        "@type" : {
          "type" : "string",
          "enum" : [ "createStreamV1",  "createStreamV2" ],
          "default" : "createStreamV1"
        },
        "sourceName" : {
          "type" : "string"
        },
        "schema" : {
          "type" : "string"
        },
        "keyField" : {
          "type" : "string"
        },
        "timestampColumn" : {
          "$ref" : "#/definitions/TimestampColumn"
        },
        "kafkaTopicName" : {
          "type" : "string"
        },
        "formats" : {
          "$ref" : "#/definitions/Formats"
        },
        "windowInfo" : {
          "$ref" : "#/definitions/WindowInfo"
        },
        "awesomeness" : {
          "$ref" : "#/definitions/Awesomeness"
        }
      },
      "title" : "createStreamV1",
      "required" : [ "@type", "sourceName", "kafkaTopicName", "formats" ]
    }
}
```

Note: no `schema` in `required` :( .  Plus that `title` is suspect ;)

Instead we have:

```json
{
"CreateStreamCommandV1" : {
      "type" : "object",
      "additionalProperties" : false,
      "properties" : {
        "sourceName" : {
          "type" : "string"
        },
        "schema" : {
          "type" : "string"
        },
        "keyField" : {
          "type" : "string"
        },
        "timestampColumn" : {
          "$ref" : "#/definitions/TimestampColumn"
        },
        "kafkaTopicName" : {
          "type" : "string"
        },
        "formats" : {
          "$ref" : "#/definitions/Formats"
        },
        "windowInfo" : {
          "$ref" : "#/definitions/WindowInfo"
        }
      },
      "title" : "createStreamV1",
      "required" : [ "@type", "sourceName", "schema", "kafkaTopicName", "formats" ]
    },
"CreateStreamCommandV2" : {
      "type" : "object",
      "additionalProperties" : false,
      "properties" : {
        "sourceName" : {
          "type" : "string"
        },
        "keyField" : {
          "type" : "string"
        },
        "timestampColumn" : {
          "$ref" : "#/definitions/TimestampColumn"
        },
        "kafkaTopicName" : {
          "type" : "string"
        },
        "formats" : {
          "$ref" : "#/definitions/Formats"
        },
        "windowInfo" : {
          "$ref" : "#/definitions/WindowInfo"
        },
        "awesomeness" : {
          "$ref" : "#/definitions/Awesomeness"
        }
      },
      "title" : "createStreamV1",
      "required" : [ "@type", "sourceName", "awesomeness", "kafkaTopicName", "formats" ]
    }
}
```

So each step type stands alone and is strongly typed / validated.  To put it another way, I don't think we need step types to be polymorphic.  `CreateStreamCommandV2` should be distinct from `CreateStreamCommandV1`. They don't need to share a common interface. In fact doing so massively restricts what we can change.

The same is also true of other types, e.g. the `KsqlPlan` and `ExecutionStepProperties`.  

For `ExecutionStepProperties` I don't see why we'd need the ability to swap in different property impls. If you can think of a reason then an example would be magic.  As it is, I think we can simplify the execution steps so that their property type is concretely `DefaultExecutionStepProperties`, i.e. 

```json
{
"properties" : {
          "oneOf" : [ {
            "$ref" : "#/definitions/DefaultExecutionStepProperties"
          } ]
        }
}
```

becomes 

```json
{
"properties" : {
        "$ref" : "#/definitions/DefaultExecutionStepProperties"
     }
}
```

We can add new properties to the properties holder if we need.  If we really really needed polymorphic properties then we would be able to just swap the schema back to the `oneOf` format, as all existing JSON plans would still match this schema

Reply:
> For ExecutionStepProperties I don't see why we'd need the ability to swap in different property impls. If you can think of a reason then an example would be magic. As it is, I think we can simplify the execution steps so that their property type is concretely DefaultExecutionStepProperties, i.e.

Yes, I agree here. Cleaning that up.

Reply:
> So we need the @type for the Jackson stuff to work? But only if we need the type to be polymorphic, right? And I don't think it does.

It's about restricting the type when it's referenced , not about the type being polymorphic so that properties can be shared.

For example, most steps have one or more source steps:
```
"StreamSelect" : {
    ...
    "source": <schema>
    ...
}
```
I want <schema> to be `oneOf` a list of possible source step schemas rather than just `"object"`. But to do that we need some way to identify which schema in the list to validate against when validating json data. That's what the type enum is for - it serves as a type identifier in the json payload. 

So in the create stream example, you'd have:
```
{
"CreateStreamCommandV1" : {
      "type" : "object",
      "additionalProperties" : false,
      "properties" : {
        "@type" : {
       "type" : "string",
       "enum" : [ "createStreamV1" ],
    },
        ... 
},
"CreateStreamCommandV2" : {
      "type" : "object",
      "additionalProperties" : false,
      "properties" : {
        "@type" : {
       "type" : "string",
       "enum" : [ "createStreamV2" ],
    },
    ...
      }
}
```